draft/ZEP0002.md

```diff
@@ -0,0 +1,285 @@
+in a backwards compatible manner.
+
+
+## Detailed description
+
+
+### Combining regular blocks of chunks in a shard
+
+To add sharding to Zarr, data that was stored previously as separate objects must be combined.
+Different strategies can be considered when sharding array data, such as:
+1. sharding across single dimensions,
+2. sharding by regular n-dimensional blocks of data, and
+3. sharding random combinations of chunks via hashing.
+
+All of those strategies only shard the array-data, both as metadata must first be read without
+sharding, so that sharding itself can be configured in the metadata, and also since the relevant
+parts of the data is the array data itself.
+
+Strategy 2 is a strict super-set of strategy 1, as sharding across a single dimension can
+be produced by sharding with n-dimensional blocks where all blocks consist of single chunks
+except for the selected dimension.
+
+Strategy 3 allows to balance loads better between shards than the other two strategies. However,
+sharding is not meant to distribute loads in this case, since all shards will mainly be served
+via the same underlying storage, but it instead reduces the number of stored files or objects.
+
+Comparing strategy 3 and 2, in opposite to the load balancing, it is more important to co-locate
+data that us usually accessed together or in short succession. This is optimized by combining
+multiple chunks that are close to each other, what is supported best by strategy 2, using
+regular n-dimensional blocks of chunks as sharding units.
+
+
+### Shardings as a storage transformer extenion
+
+Sharding can be an optional addition to the current Zarr features, as it is primarily important for
+very large arrays. Therefore it should be added as an extension, which can enhance different Zarr
+implementations.
+
+Sharding by combining different chunks can only be implemented in logical units operating after
+chunking and indexing is done. Therefore could be implemented directly in the storage layer,
+but this could result in different sharding implementations between different stores and Zarr
+implementations, compromising the core Zarr benefit of interoperability.
+
+Therefore, sharding should rather be implemented as a storage transformer, which allows
+to consistently adress array data across different stores and ensures compatibility
+across implementations, by specifying an authoritative specification.
+
+
+### Configuration
+
+Sharding can be configured per array in the array metadata by specifiying the
+following array metadata keys:
+* `extension` must always be `"https://purl.org/zarr/spec/storage_transformers/sharding/1.0"`
+  for this extension.
+* `type` specifies a binary shard format. In this version, the only binary format is the
+  `indexed` format.
+* `configuration` contains only the following configuration key:
+  * `chunks_per_shard` is an array of integers providing the number of chunks that are combined in
```

Comment:
What about border shards? I think we need to write somewhere how they are treated. (I assume the same index size as all other shards, but with a bunch of empty (`2^64 - 1`) chunk entries outside the border).

Reply:
It's written in the `Binary shard format` section:
> he index always has the full shape of all possible chunks per shard, even if they are outside of the array size.

Do you think it should be part of the configuration section as well?